### PR TITLE
feat: team update sync

### PR DESCRIPTION
**Pull request by ana-yldio**
https://trello.com/c/wQYMzkB8/1568-3-algolia-is-not-receiving-updates-regarding-the-teams-of-a-shared-output-must

### CODE REVIEW

File: apps/asap-server/src/handlers/teams/index-handler.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    event: EventBridgeEvent<TeamsEventType, SquidexWebhookTeamPayload>,
+  ) => Promise<void>) =>
+  async (
+    event: EventBridgeEvent<TeamsEventType, SquidexWebhookTeamPayload>,
+  ): Promise<void> => {
+    const team = await teamController.fetchById(event.detail.payload.id);
+
+    logger.info(`Fetched team with id ${team.id}`);
+
+    await indexResearchOutputs(team, algoliaClient);
```
Comment (peterstarling) [inline apps/asap-server/src/handlers/teams/index-handler.ts:27]: I see you tested the handler without mocking this so it's not really super relevant but... 

you could make the `indexResearchOutput` a function that takes a `ResearchOutputResponse` type and use it here and in [this handler](https://github.com/yldio/asap-hub/blob/master/apps/asap-server/src/handlers/research-output/index-handler.ts). Technically that would abstract away the algolia implementation and would enable you to mock it (which may become handy should you need to test some edge cases here).

Comment (peterstarling) [reply apps/asap-server/src/handlers/teams/index-handler.ts:27]: one more thing - your `indexResearchOutputs` takes a dependency that it inherits from `indexHandlerFactory` - thats not a good pattern, in general, [see here](https://en.wikipedia.org/wiki/Dependency_inversion_principle)

`indexHandlerFactory` is not even interested in `algoliaClient`, so realistically it should instead take a `indexResearchOutput` dependency instead and not care about what that function does.

Comment (peterstarling) [reply apps/asap-server/src/handlers/teams/index-handler.ts:27]: finally :D you could just leave it as is because your `indexResearchOutputs` is just a wrapper around some logic - it's just strange to see that you're having to pass down the algoliaClient - I guess since your `indexHandlerFactory` is a factory you could add the init phase to it (before returning the function) and define your `indexResearchOutputs` there - that way `algoliaClient` will already be available in the scope

File: apps/asap-server/test/handlers/teams/index-handler.test.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  afterEach(() => jest.clearAllMocks());
+
+  test('Should fetch the team and create a record in Algolia for every research output', async () => {
+    teamControllerMock.fetchById.mockResolvedValueOnce(teamResponse);
+
+    await indexHandler(getEvent());
+
+    expect(algoliaIndexMock.saveObjects).toHaveBeenCalledWith(
+      teamResponse.outputs.map((output) => ({
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/teams/index-handler.test.ts:29]: not a huge fan of such logic in tests -especially since you are effectively coping it from the implementing function. 

I would probably reassign `teamResponse.outputs` in the test and then expect these exact objects that you define.

File: apps/asap-server/test/handlers/teams/index-handler.test.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+describe('Team Research Outputs Index', () => {
+  const indexHandler = indexHandlerFactory(
+    teamControllerMock,
+    algoliaClientMock,
+  );
+
+  afterEach(() => jest.clearAllMocks());
+
+  test('Should fetch the team and create a record in Algolia for every research output', async () => {
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/teams/index-handler.test.ts:23]: maybe add a test that confirms that algolia client is not called when the team has no associated researchOutputs?

File: apps/asap-server/test/handlers/webhooks/webhook-teams.test.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          EventBusName: eventBus,
+          Source: eventSource,
+          DetailType: 'TeamsUpdated',
+          Detail: JSON.stringify(updateTeamEvent()),
+        },
+      ],
+    });
+  });
+
+  test('Should put the teams-updated event into the event bus and return 200', async () => {
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/webhooks/webhook-teams.test.ts:81]: its the same test as the above, no?

File: packages/squidex/schema/rules/rule4.json [hunk 5]
```diff
@@ -1,13 +1,13 @@
 {
   "$schema": "./../__json/rule.json",
   "isEnabled": true,
-  "name": "Teams upated",
+  "name": "Teams Updated",
```
Comment (peterstarling) [inline packages/squidex/schema/rules/rule4.json:4]: hang on - did we have this rule already? i see it was added as part of [this PR](https://github.com/yldio/asap-hub/pull/1071/files#diff-c7331d70cf82ee31cdc53c03fd9f2e587bc72de2156c271e43c3083588ae16b7) - i take it it was added accidentally? how did it slip through the PR?

File: packages/squidex/schema/rules/rule4.json [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
   "trigger": {
     "triggerType": "ContentChanged",
     "schemas": [
       {
         "schemaId": "teams",
-        "condition": "event.type == 'Created' || event.type == 'Updated'"
+        "condition": "(event.type == 'Updated' || event.type == 'Published') && event.status == 'Published'"
```
Comment (peterstarling) [inline packages/squidex/schema/rules/rule4.json:10]: see above comment - basically if we know for sure that the rule was here accidentally you are free to change it but otherwise i would check why it was here in the first place

File: packages/squidex/schema/rules/rule4.json (hunk 5 above)
Comment (ana-yldio) [reply packages/squidex/schema/rules/rule4.json:4]: yes, it was a mistake

File: apps/asap-server/src/handlers/teams/index-handler.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import { TeamsEventType } from '../webhooks/webhook-teams';
+import Team, { TeamController } from '../../controllers/teams';
+import {
+  algoliaAppId,
+  algoliaIndexApiKey,
+  algoliaResearchOutputIndex,
+} from '../../config';
+import logger from '../../utils/logger';
+
+export const indexHandlerFactory = (
```
Comment (peterstarling) [inline apps/asap-server/src/handlers/teams/index-handler.ts:12]: I'd rename this one to `indexResearchOutputtByTeamHandler` and the other one to `indexResearchOutputHandler` or something along these lines cause it will be easier to use the named import this way

File: apps/asap-server/test/handlers/teams/index-handler.test.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      objectID: output.id,
+    }));
+
+    await indexHandler(getEvent());
+    expect(algoliaIndexMock.saveObjects).toHaveBeenCalledWith(
+      teamOutputsIndexes,
+    );
+  });
+
+  test('Should not trigger algolia when there are no research outputs', async () => {
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/teams/index-handler.test.ts:42]: `should not trigger algolia save when there are no research outputs associated with the team`

File: apps/asap-server/test/handlers/teams/index-handler.test.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      },
+    ];
+
+    outputs.forEach((output) =>
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce(output),
+    );
+
+    const updateEvent = getEvent();
+    updateEvent.detail.payload.data.outputs.iv = outputs.map((ro) => ro.id);
+    await indexHandler(getEvent());
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/teams/index-handler.test.ts:41]: you're not actually using the `updateEvent` from above, im surprised it still works

File: apps/asap-server/test/handlers/teams/index-handler.test.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+    expect(algoliaIndexMock.saveObject).toHaveBeenCalledTimes(2);
+    expect(algoliaIndexMock.saveObject).toHaveBeenNthCalledWith(1, {
```
Comment (peterstarling) [inline apps/asap-server/test/handlers/teams/index-handler.test.ts:44]: we dont really care about the order of calls do we? i would simply assert that the call was made

**Comment by ana-yldio** [comment]
move to https://github.com/yldio/asap-hub/pull/1086
exceed api calls limit
